### DISCUSSION TITLE
unit-testing services that use bullMQ with redis

### DISCUSSION THREAD

**Discussion by RafiuJaman** [taskforcesh/bullmq#2091] [Q&A] [answered]
Hi guys!, hope you are all doing well.

i have found myself in a pinch.
we have a monolith which uses bull mq for scheduling jobs such as emials notifs etc,

so far i have been running redis in a docker container for unit tests but i realise this not a unit-test everything in a unittest should happen in memory and i am unable to find a solution that helps me.
after removing he docker and runnning the testsi basically get this

```
  at Queue.emit (node_modules/bullmq/src/classes/queue-base.ts:110:17)
      at Queue.emit (node_modules/bullmq/src/classes/queue.ts:138:18)
      at RedisConnection.<anonymous> (node_modules/bullmq/src/classes/queue-base.ts:61:56)
      at EventEmitter.RedisConnection.handleClientError (node_modules/bullmq/src/classes/redis-connection.ts:101:12)

  console.error
    Error: connect ECONNREFUSED 127.0.0.1:6379
        at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1494:16) {
      errno: -61,
      code: 'ECONNREFUSED',
      syscall: 'connect',
      address: '127.0.0.1',
      port: 6379
    }
  
```
    
    my guess is that i need to run an in memory redis similar to how an sqllitedb works in unit tests
    
    i think i cannot use bullmq mocks as the services are nested and some use the queue and some dont.
    
    for example i am writing unittests for foe service1 but service 1 imports services2 and service2 (in the module file has the queue initialised)
    

```
    @Module({
  imports: [
    BullModule.forRoot({
      /**
       * Redis client connection string
       */
      connection: {
        host: REDIS_HOST,
        port: parseInt(REDIS_PORT),
        password: REDIS_PASSWORD,
      },
    }),
  ],
  controllers: [],
  providers: [QueueJobsRepository, QueueService],
  exports: [BullModule, QueueService],
})
export class QueueModule {}

```

```
@Module({
  imports: [
    ConfigModule,
    QueueModule,
    // more queues can be added here
    BullModule.registerQueue({
      name: EMAIL_QUEUE,
    }),
  ],
  exports: [EmailService],
  providers: [EmailService, EmailConsumer, EmailQueueService],
  controllers: [],
})
export class EmailModule {}
```

any help is greatly appriciated

**Comment by manast** [comment]
We recommend to use a Redis instance in order to run the tests.

**Comment by RafiuJaman** [comment]
i did try doing that, have tried ioredis, redis-mock, redis-server none of these helped,
can you tell me a library that you guys have used with some code?

**Comment by roggervalf** [comment]
hi @RafiuJaman, we use an actual redis instance in our local or also we have a docker compose to instantiate one https://github.com/taskforcesh/bullmq/blob/master/docker-compose.yml

**Reply by RafiuJaman** [reply]
Hi @roggervalf  i had done the exact same thing but this surely does not constitute a unit test right?
it is more like an integration test or e2e

**Reply by manast** [reply]
Yes, it would be an integration test, if you want to implement unit tests then you would need to mock BullMQ in your tests.

**Reply by RafiuJaman** [reply]
any examples for that?
@manast

**Reply by manast** [reply, marked as answer]
This would be specific for your unit tests, like you would mock any other external library.
